beetsplug/bareasc.py

```diff
@@ -0,0 +1,83 @@
+# -*- coding: utf-8 -*-
+# This file is part of beets.
+# Copyright 2016, Philippe Mongeau.
+# Copyright 2021, Graham R. Cobb.
+#
+# Permission is hereby granted, free of charge, to any person obtaining
+# a copy of this software and ascociated documentation files (the
+# "Software"), to deal in the Software without restriction, including
+# without limitation the rights to use, copy, modify, merge, publish,
+# distribute, sublicense, and/or sell copies of the Software, and to
+# permit persons to whom the Software is furnished to do so, subject to
+# the following conditions:
+#
+# The above copyright notice and this permission notice shall be
+# included in all copies or substantial portions of the Software.
+#
+# This module is adapted from Fuzzy in accordance to the licence of
+# that module
+
+"""Provides a bare-ASCII matching query."""
+
+from __future__ import division, absolute_import, print_function
+
+from beets import ui
+from beets.ui import print_, decargs
+from beets.plugins import BeetsPlugin
+from beets.dbcore.query import StringFieldQuery
+from unidecode import unidecode
+
+
+class BareascQuery(StringFieldQuery):
+    """Compare items using bare ASCII, without accents etc."""
+    @classmethod
+    def string_match(cls, pattern, val):
+        """Convert both pattern and string to plain ASCII before matching.
+
+        If pattern is all lower case, also convert string to lower case so
+        match is also case insensitive
+        """
+        # smartcase
+        if pattern.islower():
+            val = val.lower()
+        pattern = unidecode(pattern)
+        val = unidecode(val)
+        return pattern in val
+
+
+class BareascPlugin(BeetsPlugin):
+    """Plugin to provide bare-ASCII option for beets matching."""
+    def __init__(self):
+        """Default prefix for selecting bare-ASCII matching is #."""
+        super(BareascPlugin, self).__init__()
+        self.config.add({
+            'prefix': '#',
+        })
+
+    def queries(self):
+        """Register bare-ASCII matching."""
+        prefix = self.config['prefix'].as_str()
+        return {prefix: BareascQuery}
+
+    def commands(self):
+        """Add bareasc command as unidecode version of 'list'."""
+        cmd = ui.Subcommand('bareasc',
+                            help='unidecode version of beet list command')
+        cmd.parser.usage += u"\n" \
+            u'Example: %prog -f \'$album: $title\' artist:beatles'
+        cmd.parser.add_all_common_options()
+        cmd.func = self.unidecode_list
+        return [cmd]
+
+    def unidecode_list(self, lib, opts, args):
+        """Emulate normal 'list' command but with unidecode output."""
+        query = decargs(args)
+        album = opts.album
+        fmt = u''
```

Comment:
Is this intentional? Maybe it's supposed to get `fmt` from the options?

Reply:
Hmm. Actually I was confused, and never went back to ask for help.

I took the code from commands.py: list_func and list_items.  `list_func` has three args: `(lib, opts, arg)` -- I understand those, I think. It then calls `list_items` with `(lib, decargs(args), opts.album)` -- again I think I understand that. However, `list_items` is declared with a fourth optional argument `fmt=u''` which it passes to `format`. It seems that in this use case (handling the `beet list` command itself), fmt is not being specified so I wanted to make sure the call from the `bareasc` command gave identical results by explicitly calling `format` with second argument `''`.

I am not familiar enough with the python `format` function to know if setting the second arg to '' is the same as dropping it completely. If it is, I can just remove `fmt`. Or is there something else I should be doing?

Reply:
If we're just calling `format` to invoke `__str__`, we should probably just do `str(item)` (not sure if this will cause any encoding issues though?).

Reply:
Yes, that makes much more sense to me. And it seems to work fine (at least in Python3). I will push that change, along with a test case for it, and we can check if it works for Python2 as well.

Reply:
Tests pass. Thanks for the help Adrian and Jack. I assume you have no problem with me marking this as resolved.

Reply:
Awesome! Thanks to both of you. Of course; seems resolved to me. :smiley: